### PR TITLE
Allow quantize to only copy tensors, other improvements

### PR DESCRIPTION
**Pull request by KerfuffleV2**
You can now specify `copy` as the type to the `quantize` tool to only copy tensors, never quantize. I forget where I saw it, but it was recently mentioned that something like could be used to convert/repackage from stuff like GGUF V1 to GGUF V2.

This also improves the logic in the actual quantize function (mainly for k-quants) to be a bit smarter when requantizing to the same format. The current logic skips quantizing when `quantize_type == tensor->type`. However, this means the k-quants logic for stuff like counting numbers of tensors to decide whether to use more bits doesn't run. The result is if you quantize to `q4_k_m` and then requantize to `q4_k_m` stuff like requantizing `q6_k` tensors to `q4_k` will occur. Not super likely, but I recall reading a pull recently where someone was confused by behavior like that and making it work a little more intuitively is pretty easy. (Of course, if the k-quants strategy changes then you're still going to have a bad time but this change is at least as good as the status quo.)

### CODE REVIEW

**Comment by JackJollimore** [comment]
Related: https://github.com/ggerganov/llama.cpp/pull/2821#issuecomment-1694511206

File: examples/quantize/quantize.cpp [hunk 1]
```diff
@@ -76,7 +78,12 @@ void usage(const char * executable) {
     fprintf(stderr, "  --leave-output-tensor: Will leave output.weight un(re)quantized. Increases model size but may also increase quality, especially when requantizing\n");
     fprintf(stderr, "\nAllowed quantization types:\n");
```
Comment (ggerganov) [inline examples/quantize/quantize.cpp:79]: I think these `fprintf` should become `printf` so that help goes to `stdout`

**Review by ggerganov — APPROVED** [review]

File: examples/quantize/quantize.cpp (hunk 1 above)
Comment (KerfuffleV2) [reply examples/quantize/quantize.cpp]: Should stuff like that tool be adapted to use the logging system? I see the internal quantize stuff in `llama.cpp` reports its progress with `LLAMA_LOG_INFO`, etc.

Comment (ggerganov) [reply examples/quantize/quantize.cpp]: I don't think so - it's specific to the current tool. `llama.cpp` can be used by 3rd party projects and that's why it needs the logging interface

**Pull request merged**